pilot/pkg/serviceregistry/kube/controller/ambient/multicluster.go

```diff
@@ -0,0 +1,549 @@
+		filter,
+		"EndpointSlices",
+		opts,
+		nil,
+	), opts.WithName("AllEndpointSlices")...)
+	endointSliceInformersByCluster := informerIndexByCluster(allEndpointSlices)
+	globalEndpointSlices := krt.NewCollection(clusters, collectionFromCluster[*discovery.EndpointSlice]("EndpointSlice", endointSliceInformersByCluster, opts))
+	globalEndpointSlicesByCluster := nestedCollectionIndexByCluster(globalEndpointSlices)
+
+	allNodes := krt.NewCollection(clusters, informerForCluster[*v1.Node](
+		filter,
+		"Nodes",
+		opts,
+		func(c Cluster) kclient.Filter {
+			return kclient.Filter{
+				ObjectFilter:    c.Client.ObjectFilter(),
+				ObjectTransform: kubeclient.StripNodeUnusedFields,
+			}
+		},
+	), opts.WithName("AllNodes")...)
+	nodeInformersByCluster := informerIndexByCluster(allNodes)
+	globalNodes := krt.NewCollection(clusters, collectionFromCluster[*v1.Node]("Nodes", nodeInformersByCluster, opts))
+	// Set up collections for remote clusters
+	GlobalNetworks := buildGlobalNetworkCollections(
+		clusters,
+		globalGatewaysByCluster,
+		options,
+		opts,
+	)
+	a.networks = GlobalNetworks
+	GlobalWaypoints := GlobalWaypointsCollection(
+		clusters,
+		globalGatewaysByCluster,
+		globalGatewayClassesByCluster,
+		globalPodsByCluster,
+		GlobalNetworks.SystemNamespaceNetworkByCluster,
+		opts,
+	)
+	// We need to get the waypoints for the local cluster, so we need to wait for the
+	// collection to be synced
+	if !GlobalWaypoints.WaitUntilSynced(opts.Stop()) {
+		return fmt.Errorf("could not sync global waypoints collection")
+	}
+	WaypointsPerCluster := nestedCollectionIndexByCluster(GlobalWaypoints)
+	waypointCollections := WaypointsPerCluster.Lookup(options.ClusterID)
+	if len(waypointCollections) == 0 {
+		return fmt.Errorf("could not find waypoint collection for local cluster %s", options.ClusterID)
+	}
+	LocalWaypointsWithCluster := waypointCollections[0]
+	LocalWaypoints := krt.MapCollection(LocalWaypointsWithCluster, func(obj config.ObjectWithCluster[Waypoint]) Waypoint {
+		if obj.Object == nil {
+			return Waypoint{}
+		}
+		return *obj.Object
+	})
+	// AllPolicies includes peer-authentication converted policies
+	AuthorizationPolicies, AllPolicies := PolicyCollections(LocalAuthzPolicies, LocalPeerAuths, LocalMeshConfig, LocalWaypoints, opts, a.Flags)
+	AllPolicies.RegisterBatch(PushXds(a.XDSUpdater,
+		func(i model.WorkloadAuthorization) model.ConfigKey {
+			if i.Authorization == nil {
+				return model.ConfigKey{} // nop, filter this out
+			}
+			return model.ConfigKey{Kind: kind.AuthorizationPolicy, Name: i.Authorization.Name, Namespace: i.Authorization.Namespace}
+		}), false)
+
+	// Now we get to collections where we can actually merge duplicate keys, so we can use nested collections
+	GlobalWorkloadServices := GlobalMergedServicesCollection(
+		clusters,
+		globalServicesByCluster,
+		globalServiceEntriesByCluster,
+		WaypointsPerCluster,
```

Comment:
I think we need some discussion around how we handle waypoint association for remote workloads/services and the flow of traffic from point A to B for the various topologies (local waypoint, remote waypoint, both waypoints, ...)

Reply:
Agreed, especially after you reminded me that we don't read remote service entries. I'll add an agenda for the WG meeting

Reply:
I think as of this week, we came to a consensus on the traffic path (i.e. load balancing amongst some waypoint + endpoint weight heuristic). With that in mind, I think the current implementation is right as written (or at least close)